common/storage.go

```diff
@@ -0,0 +1,178 @@
+				WithTimeout(wait.DefaultTimeout).
+				WithPolling(wait.DefaultInterval).
+				Should(Succeed())
+		})
+
+		It("has created a pod with a pvc and the pvc is bound", func() {
+			Eventually(wait.Consistent(createPodWithPVC(wcClient), 10, time.Second)).
+				WithTimeout(wait.DefaultTimeout).
+				WithPolling(wait.DefaultInterval).
+				Should(Succeed())
+		})
+
+		It("has deleted a pod with a pvc", func() {
+			Eventually(wait.Consistent(deletePodWithPVC(wcClient), 10, time.Second)).
+				WithTimeout(wait.DefaultTimeout).
+				WithPolling(wait.DefaultInterval).
+				Should(Succeed())
+		})
+	})
+}
+
+func checkStorageClassExists(wcClient *client.Client) func() error {
+	return func() error {
+		// ensure we have at least one storage class available
+		storageClasses := &storagev1.StorageClassList{}
+		err := wcClient.List(context.Background(), storageClasses)
+		if err != nil {
+			return err
+		}
+		if len(storageClasses.Items) == 0 {
+			return fmt.Errorf("no storage classes found")
+		}
+
+		return nil
+	}
+}
+
+func createPodWithPVC(wcClient *client.Client) func() error {
+	return func() error {
+		pvc := &corev1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "test-pvc",
+			},
+			Spec: corev1.PersistentVolumeClaimSpec{
+				AccessModes: []corev1.PersistentVolumeAccessMode{
+					corev1.ReadWriteOnce,
+				},
+				Resources: corev1.ResourceRequirements{
+					Requests: corev1.ResourceList{
+						corev1.ResourceStorage: resource.MustParse("1Gi"),
+					},
+				},
+			},
+		}
+
+		err := wcClient.Create(context.Background(), pvc)
+		if err != nil {
+			return err
+		}
+
+		pod := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pvc-test-pod",
+			},
+			Spec: corev1.PodSpec{
+				Containers: []corev1.Container{
+					{
+						Name:  "pvc-test-container",
+						Image: "nginx",
+						VolumeMounts: []corev1.VolumeMount{
+							{
+								Name:      "test-volume",
+								MountPath: "/data",
+							},
+						},
+					},
+				},
+				Volumes: []corev1.Volume{
+					{
+						Name: "test-volume",
+						VolumeSource: corev1.VolumeSource{
+							PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+								ClaimName: "test-pvc",
+							},
+						},
+					},
+				},
+			},
+		}
+
+		err = wcClient.Create(context.Background(), pod)
+		if err != nil {
+			// if pod creation fails, delete the PVC to avoid leaving a dangling PVC
+			if deleteErr := wcClient.Delete(context.Background(), pvc); deleteErr != nil {
+				return fmt.Errorf("failed to delete PVC after Pod creation failed: %v", deleteErr)
+			}
+			return err
+		}
+
+		return nil
+	}
+}
+
+func deletePodWithPVC(wcClient *client.Client) func() error {
+	return func() error {
+
+		pod := &corev1.Pod{}
+		err := wcClient.Get(context.Background(), types.NamespacedName{Name: pvcPodName, Namespace: corev1.NamespaceDefault}, pod)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return nil
```

Comment:
We should always try and delete both here instead. Also there's no point in getting the pod or PVC before deleting. IIRC `Delete` only needs the type and name/namespace. We can ignore the `IsNotFound` error for the delete and then move on to the PVC.